atomic_reactor/pulp_util.py

```diff
@@ -0,0 +1,179 @@
+                self.log.info("SOURCE_SECRET_PATH=%s from environment", path)
+
+            # Work out the pathnames for the certificate/key pair.
+            cer = os.path.join(path, self.CER)
+            key = os.path.join(path, self.KEY)
+
+            if not os.path.exists(cer):
+                raise RuntimeError("Certificate does not exist.")
+            if not os.path.exists(key):
+                raise RuntimeError("Key does not exist.")
+
+            # Tell dockpulp.
+            self.p.set_certs(cer, key)
+
+    def _create_missing_repos(self, pulp_repos, repo_prefix):
+        repos = pulp_repos.keys()
+        found_repos = self.p.getRepos(repos, fields=["id"])
+        found_repo_ids = [repo["id"] for repo in found_repos]
+
+        missing_repos = set(repos) - set(found_repo_ids)
+        self.log.info("Missing repos: %s" % ", ".join(missing_repos))
+        for repo in missing_repos:
+            self.p.createRepo(repo, None,
+                              registry_id=pulp_repos[repo].registry_id,
+                              prefix_with=repo_prefix)
+
+    def get_tar_metadata(self, tarfile):
+        metadata = dockpulp.imgutils.get_metadata(tarfile)
+        pulp_md = dockpulp.imgutils.get_metadata_pulp(metadata)
+        layers = pulp_md.keys()
+        top_layer = dockpulp.imgutils.get_top_layer(pulp_md)
+
+        return top_layer, layers
+
+    def create_dockpulp_and_repos(self, image_names, repo_prefix="redhat-"):
+        self.p = dockpulp.Pulp(env=self.pulp_instance)
+        self._set_auth()
+
+        # pulp_repos is mapping from repo-ids to registry-ids and tags
+        # which should be applied to those repos, expected structure:
+        # {
+        #    "my-image": PulpRepo(registry_id="nick/my-image", tags=["v1", "latest"])
+        #    ...
+        # }
+        pulp_repos = {}
+        for image in image_names:
+            repo_id = image.pulp_repo
+            self.log.info("adding repo %s", repo_id)
+            tag = image.tag if image.tag else 'latest'
+            if repo_prefix:
+                repo_id = repo_prefix + repo_id
+
+            if repo_id in pulp_repos:
+                pulp_repos[repo_id].tags.append(tag)
+            else:
+                pulp_repos[repo_id] = PulpRepo(
+                    registry_id=image.to_str(registry=False, tag=False),
+                    tags=[tag]
+                )
+
+        self._create_missing_repos(pulp_repos, repo_prefix)
+
+        return pulp_repos
+
+    def get_image_ids_existing(self, layers):
+        return self.p.getImageIdsExist(layers)
+
+    def upload(self, filename):
+        self.p.upload(filename)
+
+    def copy(self, repo_id, layer):
+        self.p.copy(repo_id, layer)
+
+    def update_repo(self, repo_id, tag):
+        self.p.updateRepo(repo_id, tag)
+
+    def remove_image(self, repo_id, image):
+        self.p.remove(repo_id, image)
+
+    def publish(self, keys):
+        task_ids = self.p.crane(keys, wait=True)
```

Comment:
Because of dockpulp's unexpected behaviour when `len(keys) == 0` (it calls publish for every single repository), we should probably `assert keys` before calling `crane()`.